### PR TITLE
feat(mongoose): add support for nested properties with name "type"

### PR DESCRIPTION
**Pull request by ghost**
* Readded support for having nested properties in schema with name of "type" using "$type" as typeKey for mongoose schema type specification
* Default value for arrays was not working. This has been fixed
* Required option for arrays was not working. This has been fixed
* Added tests for all new cases
* Moved CI mongo and node versions (5 & 6 for mongo) (18 & 20 for node)

Related to https://github.com/HPInc/davinci/pull/177

Moongoose docs https://mongoosejs.com/docs/guide.html#typeKey

### CODE REVIEW

**Comment by adriasala124** [comment]
Hono seems to be failing the build for some reason with node 14. I tried locally with node 14.20 and seems to be fine...

**Comment by adriasala124** [comment]
Regarding the comment in #177 https://github.com/HPInc/davinci/pull/177#issuecomment-1942065026
I think the meaning of "Unfortunately, the change breaks the support for keywords on sub-schemas" refers to the "required" option for example. In arrays we were loosing the keyword before

**Comment by adriasala124** [comment]
> Regarding the comment in #177 [#177 (comment)](https://github.com/HPInc/davinci/pull/177#issuecomment-1942065026) I think the meaning of "Unfortunately, the change breaks the support for keywords on sub-schemas" refers to the "required" option for example. In arrays we were loosing the keyword before

Actually I am adding a test and seems that the required option is not working with arrays in any case, and it is always defaulting to an empty array.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,111 +1,111 @@
...
-            branches:
-              only:
-                - master
+    version: 2
+    build_and_publish:
+        jobs:
+            - checkout_and_test:
+                  matrix:
+                      parameters:
+                          mongo-version: ['4.4', '5']
```
Comment (joseandrespg) [inline .circleci/config.yml:96]: We should update this with the new versions and deprecate v4 as it is EOL https://www.mongodb.com/legal/support-policy/lifecycles

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,111 +1,111 @@
...
+                                npx codecov -f packages/http-server/coverage/lcov.info -F http-server
+                                npx codecov -f packages/http-server-express/coverage/lcov.info -F http-server-express
+                                npx codecov -f packages/reflector/coverage/lcov.info -F reflector
+                      - save_cache:
+                            key: v1-source-{{ .Environment.CIRCLE_SHA1 }}
+                            paths:
+                                - '~'
+    publish:
+        docker:
+            - image: cimg/node:16.18.1
```
Comment (joseandrespg) [inline .circleci/config.yml:58]: We should update EOL versions to new LTS' https://nodejs.org/en/about/previous-releases

Comment (adriasala124) [reply .circleci/config.yml:58]: moved to 18 and 20

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,111 +1,111 @@
...
-              only:
-                - master
+    version: 2
+    build_and_publish:
+        jobs:
+            - checkout_and_test:
+                  matrix:
+                      parameters:
+                          mongo-version: ['5', '6']
+                          docker-image: ['cimg/node:18.19.0', 'cimg/node:20.11.0']
```
Comment (joseandrespg) [inline .circleci/config.yml:97]: Do we need to pin the exact versions for the testing? I see some benefits with pulling the latest major releases here

```suggestion
                          docker-image: ['cimg/node:18', 'cimg/node:20']
```

**Review by Sergio-Barquero — APPROVED** [review]

File: .circleci/config.yml (hunk 3 above)
Comment (adriasala124) [reply .circleci/config.yml:97]: seems like it doesn't work. it needs the exact image tag to pull

**Review by pierissimo — APPROVED** [review]

**Pull request merged**